doc/loginflow.md

```diff
@@ -0,0 +1,313 @@
+                "i64": "AwoQOJOvmtzn4H9e3RXl7OzfUhIgOTQzM2Q1MmFlNWY3ZjdmN2U2NzdhZjU0YzllMTcwYTkaDgoFbG9naW4SBWxvZ2lu",
+                "s64": "jAUuaROGZqMCjIoD-BSow9GCMlxELWNjYGArgqeLK0Q"
+            },
+            "v": 3,
+            "cdata": {
+                "AwA": "A6Ve5aR-tWB-Q0M7ziC2TW_oC9RxW0PwIjQI_eVUhslCKaXcdTmTr5rc5-B_Xt0V5ezs31IttCIUNp99pEQKvgrQDDBRGbJ1SXRGp9WOt8DJpr9dZRfIp_nIG4S0LSzrLnF0TiEy4jZvFcXucApDF7GgPULqnwVX7X_l630n-ZCsTcdZP1W2vnJKsUc_po0jYemzyUBDVckiJz1hDgo"
+            },
+            "ns": "std:"
+        },
+        "discharge-required-error": "invalid login macaroon"
+    }
+}
+```
+
+## 3. Discharging the third party caveat
+
+The client must now try to discharge the caveat, which will prove that the user
+is who they claim to be.  This is done by contacting an auth server (in this
+case implemented on the juju controller) as follows.
+
+This flow is implemented in
+`github.com/go-macaroon-bakery/macaroon-bakery/httpbakery`, in the
+`Client.AcquireDischarge` method.
+
+> Q: Is this the macaroon/Candid flow?
+
+### 3.1. HTTP: First attempt to discharge macaroon (fails)
+
+
+`POST https://CONTROLLER_HOST:17070/auth/discharge`
+
+Query params:
+```
+caveat64=A6Ve5aR-tWB-Q0M7ziC2TW_oC9RxW0PwIjQI_eVUhslCKaXcdTmTr5rc5-B_Xt0V5ezs31IttCIUNp99pEQKvgrQDDBRGbJ1SXRGp9WOt8DJpr9dZRfIp_nIG4S0LSzrLnF0TiEy4jZvFcXucApDF7GgPULqnwVX7X_l630n-ZCsTcdZP1W2vnJKsUc_po0jYemzyUBDVckiJz1hDgo
+
+id64=AwA
+```
+
+> Q: even though it is a POST request, query params are used instead of form
+> data in the body.  Why is that?
+
+Response:
+```json
+{
+    "Code": "interaction required",
+    "Message": "cannot discharge: interaction required",
+    "Info": {
+        "InteractionMethods": {
+            "juju_userpass": {
+                "url": "/auth/form"
+            }
+        },
+        "VisitURL": "/auth/login?waitid=fef72f1d163d4f6d2eb01a49",
+        "WaitURL": "/auth/wait?waitid=fef72f1d163d4f6d2eb01a49"
+    }
+}
+```
+
+### 3.2. HTTP: Obtain a token to discharge macaroon
+
+The client chooses the "juju_userpass" interaction method, asks the user for
+their password on CLI and then verifies the credentials.
+
+`POST https://CONTROLLER_HOST:17070/auth/form`
+
+Body:
+```json
+{
+    "form": {
+        "password": "PASSWORD",
+        "user": "admin"
+    }
+}
+```
+
+Response:
+```json
+{
+    "token": {
+        "kind": "juju_userpass",
+        "value": "MDVhZjU1NDM3NjYxZGE5YTJkMGMxMDky"
+    }
+}
+```
+
+### 3.3. HTTP: Second attempt to discharge macaroon (succeeds)
+
+Now that the username/password has been verified, we have everything we need to
+get the discharge macaroon.
+
+`POST https://CONTROLLER_HOST:17070/auth/discharge`
+
+Query params:
+```
+caveat64=A6Ve5aR-tWB-Q0M7ziC2TW_oC9RxW0PwIjQI_eVUhslCKaXcdTmTr5rc5-B_Xt0V5ezs31IttCIUNp99pEQKvgrQDDBRGbJ1SXRGp9WOt8DJpr9dZRfIp_nIG4S0LSzrLnF0TiEy4jZvFcXucApDF7GgPULqnwVX7X_l630n-ZCsTcdZP1W2vnJKsUc_po0jYemzyUBDVckiJz1hDgo
```

Comment:
For the purpose of this doc, it would be best to decode this base64 string so show the caveat content for readability

Reply:
Piping this through `base64 --decode` gives an error, not sure what I am doing wrong here.

Reply:
Ah yeah, it's actually encoded with RawURLEncoding and the payload is the encrypted caveat based on the macaroon root key. I think it corresponds to the time before caveat "time-before 2022-03-10T09:26:13.554951585Z".

I wonder if it's of any benefit to the target audience to see this level of internal detail, or whether they would benefit more from understanding the "plain" content of what is being passed around via macaroons, ie 
time-before 2022-03-10T09:26:13.554951585Z
need-declared username admin

etc